### ISSUE TITLE
Test analysis accuracy

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/gpc-web-crawler#54]
We should check the performance of the crawler before we do the crawl. @OliverWang13 and @katehausladen will look into that.

**Comment by SebastianZimmeck** [comment]
As we discussed today, we want to check the performance of our analysis (e.g., that a Do Not Sell link is correctly identified, that the US Privacy String is correctly identified before and after sending the GPC signal, ...). To that end we [use a validation set and a test set](https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets) (we leave out the training set as we are not training an ML classifier but instead have a rule-based analysis approach).
 
Here is the copy of an analysis document that we used earlier: [GPC_Detection_Performance](https://docs.google.com/spreadsheets/d/1ymyd6iPNJNbWIhYU4q4qA9PfYvu9QthFsvd_zLLCJ2M/edit?usp=sharing). It still has the sites from the OptMeowt analysis in there. But I think we would need new sites for our test and validation sets. We can change this document as we would like, delete/add columns and values, etc.

@katehausladen, please take a look. If anything is unclear, we can clarify it here or via MS Teams. @OliverWang13, you are already familiar with this document and can also provide explanations.

The first order of business is the selection of which sites to use for the validation and test sets. It should be random but also balanced in the sense that we want to have a good number of sites with Do Not Sell links and US Privacy Strings.

**Comment by SebastianZimmeck** [comment]
@katehausladen will provide in the readme of this repo links to the new validation and test set documents (and, if necessary, some brief information on how to use them/methodology, possibly in the documents).

@katehausladen and @OliverWang13 will dig a bit deeper into the reasons why the failures occurred, fix them as far as possible, re-run on the validation set to make sure the fixes work, and run on a new validation set (possibly repeat until we have enough confidence to achieve good results in the test set).

**Comment by katehausladen** [comment]
The colab notebook and validation set files are [here](https://drive.google.com/drive/u/1/folders/1UIUV_xLhZhpGEGzYwvZhFystA7D7Tvhf). Importing the validation set files into the colab notebook happens automatically as long as you allow Google to mount drive.

**Comment by SebastianZimmeck** [comment]
@katehausladen and @OliverWang13, as a reminder, here is the methodology from our [previous paper (section 4.2.2)](https://sebastianzimmeck.de/zimmeckEtAlGPC2023.pdf):

 > We obtained ground truth data for the sites in our Validation and Test Sets by performing our analysis method manually. In detail, for each site we performed the following steps:
> 
> (1) Visual inspection for a Do Not Sell link scrolling to the end of the site as necessary
> (2) Detection of the US Privacy String value by calling the USPAPI from the Firefox Web Console and checking the site’s cookies via the Firefox Network Monitor
> (3) Sending a GPC signal to the site
> (4) Repeating the second step to detect changes in the US Privacy String, if any
> 
> We declared a US Privacy String present on a site if the USPAPI returned a US Privacy String value or we found such in a us_privacy or similarly named cookie. We did not consider a null return value from a USPAPI call as a valid US Privacy String value. As sites are only required to honor GPC for California residents we performed all our analyses while being connected to a VPN with a Los Angeles IP address.

More details are in the methodologies tab in the Google Sheet from the previous validation and testing.

Unless there is a reason to deviate, we should stay close to this original methodology.

**Comment by OliverWang13** [comment]
I was recalling this methodology earlier as well. The way Kuba and I did it before, we used the earliest version of optmeowt analysis mode where we had to manually start and stop analysis. We were using the extension to send the gpc signal and then comparing our ground truth findings with what the extension found in order to test performance. This methodology was largely developed to measure the performance of analysis mode, and may not scale quite as well for measuring the performance of a crawler. We could use this methodology for finding the ground truth still and then compare it with the findings of the crawler, or create a new methodology. I would personally be alright with using an older version of optmeowt in order to send the gpc signal and find the ground truth, then compare with the findings of the crawler. @katehausladen, do you have any opinions on this or have you been finding the ground truth in any other way?

**Comment by katehausladen** [comment]
>  I would personally be alright with using an older version of optmeowt in order to send the gpc signal and find the ground truth, then compare with the findings of the crawler.

I agree with you; this is how I'd been finding the ground truth.

**Comment by SebastianZimmeck** [comment]
> The way Kuba and I did it before, we used the earliest version of optmeowt analysis mode where we had to manually start and stop analysis. This methodology was largely developed to measure the performance of analysis mode, and may not scale quite as well for measuring the performance of a crawler. 

Yes, we no longer have the manual starting and stopping of each analysis for our crawler. Though, the size of our validation and test sets for evaluating the accuracy of the crawler will be about the same as earlier. So, we would not have any issue with scale.

> I would personally be alright with using an older version of optmeowt in order to send the gpc signal and find the ground truth, then compare with the findings of the crawler.

Yes, we can do that. We have to be really careful, though, to make sure we are not sending GPC signals when we do not want to. An alternative may be to use [Firefox](https://blog.mozilla.org/netpolicy/2021/10/28/implementing-global-privacy-control/). Brave will not work because GPC cannot be turned off. But we can also use an older version of OptMeowt.

Another point we have to make sure is that we capture for our ground truth both variants how GPC signals can be sent --- via the header and the DOM.

So, the bottom line, @OliverWang13 and @katehausladen, please work out a draft for a testing methodology, and we can finalize it in our next meeting.

**Comment by katehausladen** [comment]
Actually, I think using the Firefox settings would be easier than using an earlier version of the extension.

Do we need the DNT header to be sent as well when finding ground truth values? I know this is automatically sent by the analysis extension, but I think we would also need to turn on the privacy.donottrackheader.enabled setting if we needed to send it.

**Comment by SebastianZimmeck** [comment]
> Do we need the DNT header to be sent as well when finding ground truth values? I know this is automatically sent by the analysis extension, but I think we would also need to turn on the privacy.donottrackheader.enabled setting if we needed to send it.

This is a very good and important point. The answer is no. We do not what to send the DNT header (and we do not want to set any opt out cookies or enable other mechanisms that would facilitate our opt out). We want to test the extent to which sites opt people out based on GPC. Thus, we should not use any other opt out mechanism. Otherwise, it is not clear on which basis the opt out occurred. 

So, let's make sure that Firefox (or whatever other software we use) does not have turned on DNT etc.

**Comment by katehausladen** [comment]
In that case, I think the Firefox settings would be the way to go. In Firefox Nightly, the DNT header is off by default and is not sent when GPC is enabled. Our extension, however, always sends both DNT and GPC headers.

**Comment by katehausladen** [comment]
I retested on the validation set from before with slightly longer timeouts. The run can be found with the rest of the validation set data in [the drive](https://drive.google.com/drive/u/1/folders/1yy-psYNI6XWalhcs3GPUcD9UL1EdNxav).

Highlights:
98/100 analyzed (sites not analyzed due to site not loading)
0/98 incorrectly identified USPSs
8/98 incorrectly identified DNSLs, as I mentioned in issue #57. 

<img width="800" alt="Screenshot 2023-07-22 at 5 32 43 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/441fa8df-b394-44ab-be7b-6b700fe2ad72">

**Comment by SebastianZimmeck** [comment]
Before we can address this issue, we first need to settle on our methodologies for identifying (1) whether a site is subject to opt out requirements (#59) and (2) is respecting GPC (#60 ).

**Comment by katehausladen** [comment]
At first glance, it looks like 11/80 sites in the validation set from the paper have GPP. Is 11 sites with GPP enough? I would be happy to add an additional 20 sites that we know have GPP. This would mean we have a validation set of 100 sites, 31 of which have GPP.

**Comment by SebastianZimmeck** [comment]
I would think that 31 GPP sites is enough. So, feel free to go ahead with your plan.

For the validation set we can basically do anything we want. The question is whether we can get that distribution in the test set as well.

**Comment by katehausladen** [comment]
What should the ground truth for the urlClassification column be? The urlClassification information is currently stored in the way I described before in issue #59:
>  I decided to store the urlClassification information in an object: the keys are the tags (i.e. tracking, fingerprinting, full list [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onHeadersReceived)), and the values are a list of urls that have this tag. I trimmed the urls so that they don't contain the "https://www." or the path. This is stored in the urlClassification column.
> 
> This is what this looks like in practice: <img alt="Screenshot 2023-09-26 at 12 45 44 PM" width="700" src="https://user-images.githubusercontent.com/40359590/270733286-d19e3086-7592-4e45-8ea0-47a8d511bffe.png">

The ground truth for the urlClassification could just be the header domains for each category (i.e. like the data is stored). The challenge with this is that the particular headers aren’t exactly the same every time the site loads. Also, in our larger data analysis, we don’t really care about the individual headers. 

We’re really just using the urlClassification information to determine whether the site is subject to the CCPA, so I think it would be better to have a binary subject_to_ccpa column in the ground truths. We would manually look through the network traffic for a 3rd party site that is in the advertising, social, or fingerprinting gen category, and if we find one, record 1. The "working" value would be determined from the urlClassification column in the crawl data (i.e., if the `fingerprinting`, `tracking_ad`, `tracking_social`, `any_basic_tracking`, or `any_social_tracking` category has at least one site, then we record 1).

**Comment by SebastianZimmeck** [comment]
> What should the ground truth for the urlClassification column be? ...

I believe we clarified these points in our last meeting. If not, feel free to say otherwise, @katehausladen.

**Comment by katehausladen** [comment]
I ran the crawl and compared the values to the ground truths we collected. A lot of the crawl values had changed since we collected the ground truths (i.e., sites with 1--- then returned USPS without dashes, etc). When I manually checked the values of the sites that were different, the values were what the crawler found. For this reason, I decided to redo the ground truths in the following way:
1. Collect the ground truths for 25 sites (same methodology as before except load each site twice)
2. Immediately afterwards, run the crawler on the same set of 25 sites
3. For the sites that failed (reached error page, timed out) or collected incorrect data (because of a human check; there were 6 of these), I installed the extension on my local Nightly browser and manually loaded the sites so that the extension could collect data for them.

This way, we get data for every site and mostly eliminate the chance that the ground truth values have changed. 

Comparing those crawl results to the updated ground truths, these are the sites that failed:
1. Cnn x2 (USPS, GPP)
2. Bloomberg (USPS)
3. Business insider x2 (USPS, GPP)
4. Us news (USPS)
5. CNbc x2 (GPP, OptanonConsent)
6. Omaha (GPP)
7. Adobe (OptanonConsent)
8. Wired (OptanonConsent)
9. New Yorker (OptanonConsent)
10. Whiskey riff (GPP)
11. Mac rumors (GPP)
12. That low carb life (GPP)
13. Celebrity net worth (GPP)
14. Healthy recipes blog (GPP)
15. Guiding tech (GPP)
16. Kitchen fun with my 3 sons (GPP)

Examining why sites failed:
1. Sites 10-16 are the reason that I loaded the site twice for the ground truth collection. They return something like DBABzw~~BVQqAAAAAgA for the first load before GPC, then add the USPS on the second load (i.e. DBABzw~1YNY~BVQqAAAAAgA). A similar thing happens after the GPC signal is sent; it takes 2 loads for the USPS in the GPP string to update. I’ll look into what the decoded versions actually say.
2. CNN, Business Insider, US News, CNBC: The extension collected the correct values before the GPC signal was sent. It returned the same values for after GPC was sent, but the values should have changed. Interestingly, CNBC and CNN had other fields that did update correctly after GPC was sent.
3. Adobe, Wired, New Yorker: Looking manually, the OptanonConsent cookie is sometimes present, sometimes not for these sites. I don’t think this is an issue with our extension.
4. Omaha: I think the GPP string is different nearly every time I’ve loaded the site. It looks like the changes happen in the TCF sections, so I’ll look to see if the US sections are the same.
5. Bloomberg: The extension did not find a USPS after GPC when it should’ve.

We can discuss our plan going forward in tomorrow’s meeting.

**Comment by SebastianZimmeck** [comment]
Thanks, @katehausladen!

What does the "x2" in "Cnn x2" mean?

**Comment by katehausladen** [comment]
Just that it failed for 2 different fields

**Comment by katehausladen** [comment]
Sites 10-16 all seem to use AdThrive (now Raptive, but the files in the website are still named Adthrive).  Here's the creation of the GPP string in one of Adthrive's files: 
![Screenshot 2023-11-27 at 11 41 54 PM](https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/b0f98875-99df-47d2-a834-c7d3751d540f)

As a reminder, all the sites using this implementation follow the same sequence of gpp string values:
1st load before GPC: DBABzw~~BVQqAAAAAgA 
2nd load before GPC: DBABzw\~1YNY\~BVQqAAAAAgA 
1st load after GPC: DBABzw\~1YNY\~BVQqAAAAAgI
2nd load after GPC: DBABzw\~1YYY\~BVQVAAAAAgI 
(note that in the 2 loads after GPC, q -> V is the actual opt out)

I'm not exactly sure what is wrong with the implementation other than that it takes 2 loads to correctly populate the GPP string both before and after GPC.

**Comment by katehausladen** [comment]
I did write the code to have the analysis extension load each site twice before and after GPC was sent. This did fix the problem GPP sites discussed above. I’m going to commit the code, but I don’t think we should go forward with this method because 
1. We know this is AdThrive’s implementation
2. The possibility of sites failing to analyze goes significantly up. This is because the site must load 4 times (without timing out, etc) as opposed to 2 times. 
3. It doesn’t seem like loading the site twice improves the data we collect for the other sites in the validation set. To really be sure about this, I’d have to redo all the ground truths, but I don’t think it’s worth the time given the other observations.

**Comment by SebastianZimmeck** [comment]
> Sites 10-16 all seem to use AdThrive (now Raptive, but the files in the website are still named Adthrive).

Great find! 

As discussed, @katehausladen will ask at AdThridve (Raptive).

**Comment by katehausladen** [comment]
I redid the validation set ground truths again following the protocol discussed on Tuesday (i.e. check ground truths while selenium is running so that we collect ground truths and working on the same load).


Results:
I was able to collect data in this way for 90/100 sites. There were no incorrect USP string or GPP strings, and there was 1 incorrect OptanonConsent cookie:
![Screenshot 2023-11-30 at 1 41 18 PM](https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/781fe736-5eaa-4bde-82dc-801478820945)
I rechecked costco multiple times and got the same result, so I'm not sure why this is incorrect. This isn't an issue with timeouts because I used a 35 second timeout after each load, and our extension would collect the data at the end of the timeout.

The remaining 10/100 were:
- human checks detected by regex (6/11). These sites do get data, but we exclude that data because we know it's wrong.
- the site failed to load in 35 seconds twice, so data couldn't be collected (4/11). Of these, 2 were some more sophisticated human check that prevents the site from attempting to load at all. The other 2 did try to load, but seemed to get stuck on loading outside scripts. For instance, myspace got stuck on reading something:
![Screenshot 2023-11-30 at 2 25 05 PM](https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/4a653f6e-afd7-4c73-993b-adf73e62a79d) 
I think it is related to the vpn because extending timeouts didn't help, and the sites load just fine in selenium without the vpn on. 

I do want to add a few more human check regular expressions, but other than that, I think we should be good to go!

**Comment by SebastianZimmeck** [comment]
Thanks, @katehausladen! 

> The remaining 10/100 ... detected by regex (6/11)

Should this be 6/10? I understand that we have 10 failure for the 100 instances.

Do you have some concrete statistics? For example, there were 15 US Privacy Strings in the GPP String, and they were all correctly identified.

**Comment by katehausladen** [comment]
Yes, typo, sorry. I fixed it. Not yet; I'll get to that tonight.

**Comment by katehausladen** [comment]
The test set is done. Here are the results:

The crawler correctly collected data for all 100 sites. 

<img width="948" alt="Screenshot 2023-12-09 at 4 51 02 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/9c3fa7e8-335d-42e9-8f12-520882bc172f">
<img width="434" alt="Screenshot 2023-12-09 at 4 51 15 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/85d0cd15-0d13-4d62-9329-fca6bfbd773c">
<img width="941" alt="Screenshot 2023-12-09 at 4 51 54 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/86e1cd65-01c6-41a2-a04d-033ae3307a60">
<img width="934" alt="Screenshot 2023-12-09 at 4 52 12 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/23fb05b3-db3f-47a0-b395-3308a6453ec5">

Methodology:
I crawled 1500 sites and then used panda's df.sample to select 60 sites randomly from the sites that were successfully crawled. Then, I selected 40 sites that were found to have GPP (again using df.sample). These formed the 100-site test set. I had to replace 2 sites in the test set. One site (oracle.com) "loads" but just loads a plain black screen. The other had a human check error on the second load. I picked replacement sites using df.sample.

**Comment by SebastianZimmeck** [comment]
Perfect, @katehausladen! 

On the methodology:

- Did you crawl using a California VPN?
- Did you do a crawl of one site and then manually confirmed the results for that site, then moved to the next site, ...?

What is striking to me:

- With GPP we only see little opting out after GPC relative to the same situation with the USPS. The OptanonConsent cookie behaves more like the USPS.

**Comment by katehausladen** [comment]
Yes, I used a California VPN (Los Angeles).

> Did you do a crawl of one site and then manually confirmed the results for that site, then moved to the next site, ...?

I actually just extended the timeouts for long enough so that I could manually collect data while the crawler was collecting data. This way, both the ground truths and working values would be collected on the same load.

> With GPP we only see little opting out after GPC relative to the same situation with the USPS. The OptanonConsent cookie behaves more like the USPS.

Yes, I noticed that as well. I'm guessing that the GPP strings that changed after GPC (about 2/3) were just reflecting that they received GPC via the GPC subsection as opposed to actually opting out.
<img width="702" alt="Screenshot 2023-12-09 at 7 10 55 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/0e4b4c4a-eb07-4ec4-833b-6213b7ff323b">

I think it would be interesting to look at the following things once I crawl all 10,000 sites: 
1) the distribution of GPP strings, USP strings, and OptanonConsent cookie among sites (i.e., are sites with GPP more likely to also have USP strings). 
2) are sites with more than one string/cookie opting out via one string/cookie but not another? (i.e. opt out via USP string but not opt out via GPP)
